Improve Glide usage for Image Loading on Android

Redth opened the pull request:
This makes use of the Glide API's for loading into the ImageView directly instead of returning a Drawable.

Redth committed:
Make Glide load into imageview directly

jonathanpeppers commented on src/Core/src/ImageSources/UriImageSourceService/UriImageSourceService.Android.cs:
```diff
@@ -13,6 +13,29 @@ namespace Microsoft.Maui
+			{
+				try
+				{
+					Glide
+						.With(imageView.Context)
+						.Load(uriImageSource.Uri.OriginalString)
+						.Into(imageView);
```
I think this is the problem I never solved -- even back in Xamarin.Forms.

Here Glide will load the image based on the ImageView's size, and MAUI might want to size the ImageView based on the size of the image. So like a StackLayout or Grid Row/Column set to Auto, would have this image as 0x0.

Redth replied:
This seems to work though... I was testing this with some cases locally using .WaitForLayout() on that call which seems to wait until the android ImageView gets sized, so I think that avoids this?

Redth committed:
Refactor glide usage and implement listener

Redth committed:
Allow image views to autosize

Redth committed:
New test image

The old one was a png and blurry at the resizes, this one is svg to get the right density versions to make it clear we don't have a scaling issue in the code.

Redth committed:
Fix image size of Glide loaded font images

Redth commented on src/Controls/samples/Controls.Sample/Pages/Controls/ImagePage.xaml:
```diff
@@ -13,24 +13,24 @@
                     Style="{StaticResource Headline}"/>
                 <Image 
                     Source="https://aka.ms/campus.jpg"
-                    HeightRequest="200" 
```
We _can_ make requests here, but if we just leave the default, Glide will try and make the imageview bigger which seems to play well with layouts so far...

Redth commented on src/Core/src/ImageSources/Android/Glide/RequestBuilderExtensions.cs:
```diff
@@ -63,5 +63,29 @@ public bool OnResourceReady(Java.Lang.Object result, Java.Lang.Object model, ITa
 			}
 		}
+
+		internal class RequestCompleteListener : Java.Lang.Object, Bumptech.Glide.Request.IRequestListener
```
Wanted a listener that didn't send back the Drawable... maybe this isn't really necessary as we can just discard the drawable later anyway? Seemed better to not care about the drawable sooner in the flow though.

Redth commented on src/Core/src/ImageSources/FontImageSourceService/FontImageSourceService.Android.cs:
```diff
@@ -7,13 +7,58 @@
+				var glyph = fontImageSource.Glyph;
+
+				var size = FontManager.GetFontSize(fontImageSource.Font);
+				var unit = fontImageSource.FontAutoScalingEnabled ? ComplexUnitType.Sp : ComplexUnitType.Dip;
```
This uses `SP` units if we have font auto scaling enabled, otherwise just `DP`... The SP ones will scale with font accessibility settings on the device.

Redth commented on src/Core/src/ImageSources/FontImageSourceService/FontImageSourceService.Android.cs:
```diff
@@ -7,13 +7,58 @@
+					var builder = glide
+						.Load(new FontImageSourceModel(glyph, textSize, typeface, color))
+						.AddListener(listener)
+						.Override(Target.SizeOriginal, Target.SizeOriginal);
```
This tells Glide to set the imageview to the size of the original bitmap we generate - otherwise it tries to fill the device screen width space and we end up with a blurry image.

jonathanpeppers commented on src/Core/src/ImageSources/FileImageSourceService/FileImageSourceService.Android.cs:
```diff
@@ -12,6 +12,37 @@ namespace Microsoft.Maui
+				{
+					var listener = new RequestBuilderExtensions.RequestCompleteListener();
+					var glide = Glide.With(imageView.Context);
+					var builder = glide
+						.Load(fileImageSource.File, imageView.Context!)
+						.AddListener(listener);
```
I should add the benchmarks from here:

https://github.com/dotnet/maui/compare/main...jonathanpeppers:ImageBenchmarks

The problem I was seeing was that `ImageView.Set*` was so much faster than Glide:

| Method | Mean | Error | StdDev | Gen 0 | Allocated |
|------------------- |-----------:|----------:|-----------:|-------:|----------:|
| SetImageResource | 7.631 µs | 0.0165 µs | 0.0146 µs | - | - |
| SetImageDrawable | 55.096 µs | 1.2665 µs | 3.6135 µs | 0.1221 | 736 B |
| GlideWithTarget | 209.044 µs | 4.3352 µs | 12.2275 µs | 0.2441 | 1,544 B |

It could be there is just way more interop from C# to Java? The example above would many extra JNI calls.

We would potentially have to write the Glide code in Java so we can do a single interop call?

Redth replied:
Yeah that was my thought too...  we really just want to throw the filename/url/font info/stream over the fence to java, in a single interop call and then wait for the callback.

Redth replied:
@jonathanpeppers I moved most of the logic into the android aar project... Would you be able to test the timings here again?

mattleibow commented on src/Core/src/ImageSources/IImageSourceService.cs:
```diff
@@ -11,6 +11,11 @@ public interface IImageSourceService
 			Android.Content.Context context,
 			CancellationToken cancellationToken = default);
+
+		Task<IImageSourceServiceResult<Android.Graphics.Drawables.Drawable>?> LoadDrawableAsync(
```
This doesn't need to return a drawable. This is a different api with the image view.

mattleibow replied:
Ah, I see it is on the result api... Should we support a non-result one? It can be a base.

Redth committed:
Add glide ImageLoader to android aar project

Redth committed:
Use the helpers from the android aar

Redth committed:
Use autoscaling property from Font

jonathanpeppers commented on src/Core/src/ImageSources/Android/ImageLoaderCallback.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Microsoft.Maui
+{
+	internal class ImageLoaderCallback : Java.Lang.Object, IImageLoaderCallback
+	{
+		TaskCompletionSource<IImageSourceServiceResult<bool>> tcsResult = new();
+
+		public Task<IImageSourceServiceResult<bool>> Result
+			=> tcsResult.Task;
+
+		public void OnComplete(Java.Lang.Boolean? success, Java.Lang.IRunnable? dispose)
+		{
+			var s = success?.BooleanValue() ?? false;
+
+			Action? disposeWrapper = null;
+			if (dispose != null)
+				disposeWrapper = dispose.Run;
+
+			tcsResult.TrySetResult(new ImageSourceServiceResult(s, disposeWrapper));
+		}
+	}
+
+	internal class ImageLoaderDrawableCallback : Java.Lang.Object, IImageLoaderDrawableCallback
```
Is there a way these two could be the same type? You pay a cost per C# type that subclasses Java or per Java type you call from C#.

Redth replied:
Made these share a type as suggested.

jonathanpeppers reviewed (commented):
I can probably add my benchmark to this branch and get an idea of the time here:

https://github.com/dotnet/maui/compare/main...jonathanpeppers:ImageBenchmarks

But I will be FTO tomorrow and back on Thursday.

Redth committed:
Merge image callbacks into one

Redth committed:
Add some benchmark tests

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/ImageLoader.java:
```diff
@@ -0,0 +1,186 @@
+import java.io.InputStream;
+
+public class ImageLoader {
+    public static void loadFromResourceId(ImageView imageView, int resourceId, ImageLoaderCallback callback)
+    {
+        imageView.setImageResource(resourceId);
+        callback.onComplete(true, null, null);
+    }
```
Could this one not have a callback at all? Then the calling code would just assume it's synchronous and finished.

jonathanpeppers replied:
Oh if this is synchronous, you can probably delete this method and just call `ImageView`.

Redth replied:
Removed this.  I was already bypassing it but forgot one place.

So really, testing loading a file is kind of unnecessary for resource drawable files since they'll never hit this code path now, as we check for the drawable reasource id by filename and use `imageView.SetImageResourceId(..)` with it instead of calling out to this Image Loader.

It's really just files elsewhere on disk that will hit this path now I think, and fonts, and urls.

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/glide/fontimagesource/FontModelResourceDecoder.java:
```diff
@@ -0,0 +1,60 @@
+import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPoolAdapter;
+import com.bumptech.glide.load.resource.bitmap.BitmapResource;
+
+public class FontModelResourceDecoder implements ResourceDecoder<FontModel, Bitmap> {
```
If classes like this are only used from Java, try removing:
```suggestion
class FontModelResourceDecoder implements ResourceDecoder<FontModel, Bitmap> {
```
And see if that prevents us getting a C# binding for this type.

Redth replied:
We don't get bindings for these already as I've `<remove-node`'d them.

Redth committed:
More work on benchmark

Redth committed:
Remove unused helper methods

Redth committed:
Simplify callback

Redth commented on src/Core/src/ImageSources/IFontImageSource.cs:
```diff
@@ -10,5 +10,7 @@ public interface IFontImageSource : IImageSource
 
 		string Glyph { get; }
+
+		bool FontAutoScalingEnabled { get; }
```
Yes! removed, missed that in the Font.

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/ImageLoader.java:
```diff
@@ -0,0 +1,171 @@
+import java.io.File;
+import java.io.InputStream;
+
+public class ImageLoader {
```
One last thought on this...

For every Java class, we have to make a JNI `FindClass()` call. So should we move all the methods here into `ViewHelper.java`?

Then maybe `ViewHelper` isn't a generic enough name, maybe something like `JavaBridge` would make sense if we put every single method inside?

Redth committed:
Use custom glide targets instead of listeners

This allows .into to be used to start the request even in cases we have no image view target, and also means we don't need to add an extra listener to the cases where we use an image view...

Redth committed:
Benchmark uses non-resource/asset file

Redth committed:
Check if streams are empty

Redth committed:
Clean up tests

Redth committed:
Add glide model loader / factor for InputStream

Redth committed:
Refactor android platform code project

This removes the old Compatibility FormsViewGroup project and brings the FormsViewGroup class into the Core aar project, renames it MauiViewGroup.

Also refactored ViewHelpers and ImageHelpers into a single PlatformInterop class.

Finally, a few changes in this commit should fix some more tests.

Redth committed:
Fix a couple tests

Redth committed:
Update interop project

Redth committed:
Rework glide interop / image service

Redth committed:
Fix more tests

Redth committed:
Clear pending operations for image view targets

Redth committed:
Only load if we are supposed to apply the source

Original logic was a bit incorrect from reworking this code.

Redth committed:
Merge branch 'main' into glidier

Redth committed:
Fix rebase

Redth committed:
More rebase fixes

Redth committed:
Fix solution filters

Redth committed:
Throw if request is cancelled in test

Redth committed:
Do not always clear first

Redth committed:
Run callback clear on ui thread

If we don't run these on the main thread we can see: Java.Lang.IllegalStateException : You can't start or clear loads in RequestListener or Target callbacks. If you're trying to start a fallback request when a load fails, use RequestBuilder#error(RequestBuilder). Otherwise consider posting your into() or clear() calls to the main thread using a Handler instead.

Redth committed:
Invoke clear on main thread in load calls too

Redth committed:
Fix refactor

Redth committed:
Fix api call
